modules/login/php/ajaxlogin.class.inc

```diff
@@ -0,0 +1,405 @@
+        }
+        $loginData['description'] = html_entity_decode(
+            $config->getSetting('StudyDescription')
+        );
+        $loginData['title']       = $config->getSetting('title');
+        $loginData['logo']        = $config->getSetting('studylogo');
+        $values['login']          = $loginData;
+
+        // Request Account Setup.
+        $requestAccountData = array();
+        $reCAPTCHAPrivate   = $config->getSetting('reCAPTCHAPrivate');
+        $reCAPTCHAPublic    = $config->getSetting('reCAPTCHAPublic');
+        // reCAPTCHA if both private and public keys are set
+        if ($reCAPTCHAPrivate && $reCAPTCHAPublic) {
+            $requestAccountData['captcha'] = $reCAPTCHAPublic; // captcha key.
+        }
+        $requestAccountData['site'] = \Utility::getSiteList();
+        $values['requestAccount']   = $requestAccountData;
+
+        return $values;
+    }
+
+    /**
+     * Reset password.
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     */
+    function resetPassword(array $values) : array
+    {
+        // Always claim success to phishers.
+        $values['success'] = 'Password reset. '
+            . 'You should receive an email within a few minutes.';
+
+        // Create the user object.
+        $username = $values['username'];
+        $user     =& \User::singleton($username);
+
+        // Check if valid user.
+        $email = '';
+        try {
+            $email = $user->getData('Email');
+            if (empty($email) || empty($user)) {
+                $values['error'] = 'invalid';
+                return $values;
+            }
+        } catch (\Exception $exception) {
+            $values['error'] = 'invalid';
+        }
+
+        try {
+            // Check if valid email.
+            if ($user->isEmailValid()) {
+                $config = \NDB_Config::singleton();
+                // Generate a new password.
+                $password = \User::newPassword();
+                // Reset the password in the database and
+                // expire password so user must change it upon login.
+                $user->updatePassword($password, '1999-01-01');
+                // Send the user an email.
+                $msg_data['study']    = $config->getSetting('title');
+                $msg_data['url']      = $config->getSetting('url');
+                $msg_data['realname'] = $user->getData('Real_name');
+                $msg_data['password'] = $password;
+                \Email::send($email, 'lost_password.tpl', $msg_data);
+            } else {
+                $values['error'] = 'invalid';
+                error_log(
+                    $_SERVER['REMOTE_ADDR']
+                    . '(' . $_SERVER['HTTP_X_FORWARDED_FOR'] . ')'
+                    . " attempted to reset password for $username"
+                );
+            }
+        } catch(\Exception $e) {
+            error_log('Unhandled exception in reset_password.');
```

Comment:
This isn't the best... without including the exception message this will essentially just hide whatever exception occurred and provide no information to help debug.

Reply:
I'll change it to email couldn't be sent because I believe that's where the error would be from.

Reply:
A more specific error might be better though and I think you should log `$e->getMessage()`. `Exception` will catch _any_ exception that occurs in this whole call chain so this could hide problems by accident.